mreg/signals.py

```diff
@@ -306,37 +358,79 @@ def update_hosts_when_zone_is_added(sender, instance, created, **kwargs):
 @receiver(post_delete, sender=Ipaddress)
 def send_event_ip_removed_from_host(sender, instance, **kwargs):
     obj = {
-        'host': instance.host.name,
-        'ipaddress': instance.ipaddress,
-        'action': 'remove_ip_from_host',
+        "host": instance.host.name,
+        "ipaddress": instance.ipaddress,
+        "action": "remove_ip_from_host",
     }
     MQSender().send_event(obj, "host.ipaddress")
 
 
 @receiver(post_save, sender=Ipaddress)
 def send_event_ip_added_to_host(sender, instance, created, **kwargs):
     obj = {
-        'host': instance.host.name,
-        'ipaddress': instance.ipaddress,
-        'action': 'add_ip_to_host',
+        "host": instance.host.name,
+        "ipaddress": instance.ipaddress,
+        "action": "add_ip_to_host",
     }
     MQSender().send_event(obj, "host.ipaddress")
 
 
+# In case of host rename, we need to know the old name, so we
+# capture it here and store it in the instance. Note that we first
+# try to get the object in its original state from the database,
+# and if it does not exist there, we set the old name to None as
+# we are being called as part of an object creation.
+@receiver(pre_save, sender=Host)
+def capture_old_name(sender, instance, **kwargs):
+    try:
+        obj = sender.objects.get(pk=instance.pk)
+        instance._old_name = obj.name
+    except sender.DoesNotExist:
+        instance._old_name = None
+
+
+# Process host events, and send them to the message queue.
+# If the hostname itself is changed, send a host_updated
+# event with both the old and new hostname.
+# The old hostname is captured in the pre_save signal above.
+# Also note that _old_name is not a field in the model, so it
+# will not be saved to the database.
 @receiver(post_save, sender=Host)
 def send_event_host_created(sender, instance, created, **kwargs):
     if created:
         obj = {
-            'host': instance.name,
-            'action': 'host_created',
+            "host": instance.name,
+            "action": "host_created",
         }
         MQSender().send_event(obj, "host")
+    else:
+        # There are situations in Django where singals do not
+        # complete correctly, or pre_save isn't triggered at
+        # all. In these cases, _old_name will not be set, but
+        # the created boolean sent to post_save will still be
+        # false.
+        # To handle such eventualities, we don't simply assume
+        # that old_name is set during updates. The best we can
+        # do in these cases is to send a host_updated event.
+        old_name = getattr(instance, "_old_name", None)
+        if old_name is not None and old_name != instance.name:
+            obj = {
+                "old_host": old_name,
+                "new_host": instance.name,
+                "action": "host_updated",
+            }
+        else:
+            obj = {
+                "host": instance.name,
+                "action": "host_updated",
+            }
```

Comment:
- What do you think about using the same field name whether we have the old name or not? E.g. "host", and "old_host" in addition if it is not None. Might be less confusing to parse.
- How about using the event name "host_renamed" when the host has been renamed? Unless that may also happen without pre_save being called.

Reply:
1. The reason I opted out of using the same field name (ie, `host`) is that I found it ambiguous. Is `host` the new hostname, or the old hostname the recipient knows? If you are going to update an entry based on the event, you need to use `old_host` if and only if `old_host` is set, as that's the hostname you know, but in every other version of the `host_updated` event, you would use `host` as the host you know. It's not a big deal, but I felt it being inconsistent.
2. We can certainly use a new event name. That would make using `host` above easier as we can define the API for the `host_rename` event.

Reply:
1. That makes sense, thanks
2. Actually, I meant action and not event name. But on second thought let's keep the event name "host" and action "host_updated", that will allow us to extend the functionality in the future if there are other changes to hosts that we want to send events for.